src/installs/visual-studio.ts

```diff
@@ -107,3 +113,383 @@ export function compareVersions(lhs: string, rhs: string): number {
+ * Currently, there is a mismatch only between x86 and win32.
+ * For example, VS kits x86 and amd64_x86 will generate -A win32
+ */
+export const generatorPlatformFromVSArch: { [key: string]: string } = {
+    x86: 'win32'
+};
+
+// The reverse of generatorPlatformFromVSArch
+export const vsArchFromGeneratorPlatform: { [key: string]: string } = {
+    win32: 'x86'
+};
+
+/**
+ * Turns 'win32' into 'x86' for target architecture.
+ */
+export function targetArchFromGeneratorPlatform(generatorPlatform?: string) {
+    if (!generatorPlatform) {
+        return undefined;
+    }
+    return vsArchFromGeneratorPlatform[generatorPlatform] || generatorPlatform;
+}
+
+/**
+ * Create the host-target arch specification of a VS install,
+ * from the VS kit architecture (host) and generator platform (target).
+ * @param hostArch The architecture of the host toolset
+ * @param targetArch The architecture of the target
+ * @param amd64Alias Whether amd64 is preferred over x64.
+ */
+export function getHostTargetArchString(hostArch: string, targetArch?: string, amd64Alias: boolean = false): string {
+    // There are cases when we don't want to use the 'x64' alias of the 'amd64' architecture,
+    // like for older VS installs, for the VS kit names (for compatibility reasons)
+    // or for arm/arm64 specific vcvars scripts.
+    if (amd64Alias) {
+        if (hostArch === "x64") {
+            hostArch = "amd64";
+        }
+
+        if (targetArch === "x64") {
+            targetArch = "amd64";
+        }
+    }
+
+    if (!targetArch) {
+        targetArch = hostArch;
+    }
+
+    // CMake preferred generator platform requires 'win32', while vcvars are still using 'x86'.
+    // This function is called only for VS generators, so it is safe to overwrite
+    // targetArch with the vcvars naming.
+    // In case of any future new mismatches, use the vsArchFromGeneratorPlatform table
+    // instead of hard coding for win32 and x86.
+    // Currently, there is no need of a similar overwrite operation on hostArch,
+    // because CMake host target does not have the same name mismatch with VS.
+    targetArch = targetArchFromGeneratorPlatform(targetArch);
+
+    return (hostArch === targetArch) ? hostArch : `${hostArch}_${targetArch}`;
+}
+
+// Gets the MSVC toolsets installed for a given VS install.
+export async function enumerateMsvcToolsets(vsInstallRoot: string): Promise<string[] | undefined> {
+    const toolsetDir = path.join(vsInstallRoot, 'VC\\Tools\\MSVC');
+    if (await fs.exists(toolsetDir)) {
+        const dirContents = await fs.readdir(toolsetDir, { 'withFileTypes': true });
+        // Only the toolsets should be this directory (each in their own directories), but filter out anything else just in case.
+        // Sort in descending order, so if searching with a 1- or 2-component version (e.g. 14.27) we'll choose the latest version first
+        return dirContents.filter(item => item.isDirectory()).map(dir => dir.name).sort().reverse();
```

Comment:
Instead of doing a string-based sort for numbers, can you use `util.parseVersion` to parse out the numeric versions and then do the comparison based on that?  Currently, I don't think we have any problems, but this will avoid potential issues around variable digit strings appearing in the wrong place in the sorted list. (e.g. "9" > "12")